### ISSUE TITLE
QZO Fine-tuning Llama-2-7b/Llama-3.1-8b Not Converging

### ISSUE THREAD

**Issue by WangFei-2019** [maifoundations/QZO#4]
Hi sifeng/kayang:

First of all, thank you for your great work on this project. I am currently trying to compare your method with others in my research, but I’ve encountered some issues with reproducing the results.

## Issue Description:

Following the instructions in the README, I attempted to quantize the *meta-llama/Llama-2-7b-hf* model using the following command:

```bash
CUDA_VISIBLE_DEVICES=0 python quantization.py --model_path meta-llama/Llama-2-7b-hf --quant_mode gptq --quant_path quantized
```

Subsequently, I fine-tuned the model on the SST2 dataset using:
```bash
CUDA_VISIBLE_DEVICES=0 MODEL=quantized/llama-2-7b-hf-gptq-b4-g128 TASK=SST2 MODE=ft LR=1e-7 EPS=1e-3 bash scripts/qzo.sh
```

To align with the experimental setup described in your paper, I added the --train_unquantized option. However, I found that the model does not converge during training, even when I adjust the learning rate.

<img width="1243" height="648" alt="Image" src="https://github.com/user-attachments/assets/2d0be28b-5fef-424c-8249-b224c8320095" />

## Environment:

Python version: 3.12.0
PyTorch version: 2.8.0
CUDA version: 12.2

Could you please give me some guidance on how to fix this issue? Any suggestions or insights would be greatly appreciated.

**Comment by sifengshang** [comment]
Thank you for being interested in our project! To accurately reproduce the results with Llama-2-7B and Llama-3.1-8B, the unquantized model weights should NOT be trainable, i.e.,  the '--train_unquantized' option should be turned off. The entire training process lasts 20,000 steps. Can you report the accuracy of Llama-2-7B and Llama-3.1-8B when the fine-tuning is finished?

**Comment by WangFei-2019** [comment]
Thank you for your guidance and clarification!

I have conducted a fine-tuning process on the **Meta-Llama-3-8b** model for 20,000 steps with the *--train_unquantized* option turned off. On the SST dataset, the loss curve shows convergence. However, I observed that on datasets such as RTE, CB, BoolQ, and SQuAD, the loss curves do not converge.

Having reviewed your paper more carefully, I have started a more rigorous reproduction on both **Llama-2-7b** and **Meta-Llama-3.1-8b**. Today, I obtained the loss curves for **Meta-Llama-3.1-8b** across all tasks with *--train_unquantized* turned on, and unfortunately, none of the tasks are converging. In the past two hours, I turned off *--train_unquantized* and observed the first 8,000 steps, but it still does not converge.

I plan to complete the full fine-tuning process and will provide the reproduced loss curves. In the meantime, could you share your social media contact information? It would be helpful to discuss more detailed reproduction information directly.

Thank you for your continued support and assistance.

**Comment by sifengshang** [comment]
Thank you for the response! Note that the loss reported is computed from the quantized model perturbed with the noise vectors, which may not behave as expected. We will also review the released codes and reproduce the results. If it is an urgent issue for you, we may exchange our social media account through email :)

**Comment by WangFei-2019** [comment]
I understand that the random Gaussian perturbations can affect the experiments, so I have fixed the random seed to mitigate this issue. I would be happy to engage in a deeper discussion and analysis with you. Please feel free to reach out to me via email at [wang_fei2019@163.com](mailto:wang_fei2019@163.com).

**Comment by WangFei-2019** [comment]
The Fine-Tuning Results of LLaMa-2-7B on the SST2 Dataset:

<img width="1266" height="697" alt="Image" src="https://github.com/user-attachments/assets/ce2cbf68-7ea8-440a-8da9-c65a6ddd7aeb" />

The Fine-Tuning Results of LLaMa-3.1-8B on the SST2 Dataset:

<img width="1256" height="669" alt="Image" src="https://github.com/user-attachments/assets/b1892dd2-a4cc-4411-8ad9-eb33205c8063" />

**Comment by WangFei-2019** [comment]
The Fine-Tuning Results of LLaMa-2-7B on the SST2 Dataset:

<img width="1266" height="697" alt="Image" src="https://github.com/user-attachments/assets/dc4785cd-076b-4083-a2dd-9a0d6afb69dc" />

The Fine-Tuning Results of LLaMa-3.1-8B on the SST2 Dataset:

<img width="1256" height="669" alt="Image" src="https://github.com/user-attachments/assets/d393f8ec-d170-445f-9ab0-ddde1f37b3de" />

**Comment by sifengshang** [comment]
After an in-depth discussion with Mr. Wang, we together figured out that the use of the Marlin inference kernel caused the divergent phenomenon. Note that in the QZO paper, the Triton inference kernel was used across all experiments, which works fine in convergence. We will fix this issue and support the Marlin together with other CUDA-based inference kernels in the future.
